### ISSUE TITLE
Have difficulty understanding the `ActionDecoder`

### ISSUE THREAD

**Issue by xlnwel** [danijar/dreamer#4]
Hi,

Sorry to bother you again. I'm very interested in Dreamer and try to reproduce it these days. However, I get stuck at the calling method of `ActionDecoder`. Specifically, I have a hard time to understand the following code originally from [here](https://github.com/danijar/dreamer/blob/02f0210f5991c7710826ca7881f19c64a012290c/models.py#L165)

```python
      mean = self._mean_scale * tf.tanh(mean / self._mean_scale)
      std = tf.nn.softplus(std + raw_init_std) + self._min_std
      dist = tfd.Normal(mean, std)
      dist = tfd.TransformedDistribution(dist, tools.TanhBijector())
      dist = tfd.Independent(dist, 1)
      dist = tools.SampleDist(dist) 
```

I cannot get the point of this line `mean = self._mean_scale * tf.tanh(mean / self._mean_scale)`, which seems to limit `mean` to some range defined by `self._mean_scale`. But why do we do that as `TanhBijector` has already limited the output of the distribution to range `[-1, 1]`? Moreover, what's the point of `Independent` and `SampleDist`? If I understand them right, `Independent` introduces dependence between actions dimension and `SampleDist` seems to draw samples from the distribution and recompute the corresponding stats from these samples. But why do we do that as we can compute these stats directly from the `mean` and `std`?

Also, I try to reframe RSSM so that I can wrap it with `tf.keras.layers.RNN`. Here's the cell structure I come up with

<img width="307" alt="image" src="https://user-images.githubusercontent.com/10184153/76273268-6e2a6c80-62b8-11ea-9c07-93f5cddaa730.png">

where $e_t$ is the embedding and $s'_t$ is the output of the posterior. Here, I move the deterministic and stochastic state models into a single RSSM cell and compute the posterior after the RNN is expanded. I humbly think this structure may be faster as we can process all posterior at once after the RNN is expanded. However, I'm not sure if I miss something. Do you think this structure is consistent with your implementation?

Thanks in advance. I know that I asked a lot, but I sincerely hope you could take some time to help me.

Best,
Sherwin Chen

**Comment by xlnwel** [comment]
After implementing my version of RSSM, I found that I misunderstood it before: RSSM is expanded differently at dynamics learning and behavior learning. I think the correct diagram should be the following one

<img width="668" alt="Screen Shot 2020-03-11 at 23 10 41" src="https://user-images.githubusercontent.com/10184153/76432199-9bc50200-63ed-11ea-8851-038ac7e2c036.png">

At the time of dynamics learning, the hidden state is computed from the output of the posterior $q$(the left part) while during the imagine phase, the hidden state is computed from the output of the prior $p$(the right part). I can see that the left matches the VAE structure. However, I can't explain why it should be preferred compared to the one I previously considered. Could you help me clarify it?

**Comment by danijar** [comment]
Hi @xlnwel, thanks for your questions. Could you please move your question about the RSSM into a second Github issue? That way, it'll be easier for others who might have similar questions.

Regarding the action distribution, the reason the mean is bounded to `[-5, +5]` before being transformed by the tanh is to avoid numerical instabilities. For computing log-probabilities, we need to invert the tanh and this becomes difficult in highly saturated regions. By the way, SAC also bounds the mean using tanh but doesn't scale up by 5 afterwards, which limits its action range a bit too much. Check out the link in the code to play around with the tanh normal distribution: https://www.desmos.com/calculator/rcmcf5jwe7

I'm closing the issue for now, but please follow up if this doesn't answer your question.

**Comment by xlnwel** [comment]
Hi @danijar, 

Thanks for the explanation. Now I see why you limit the mean. Also, I've done some experiments on my own and learned that `Independent` affects the way `log_prob` is computed but not the way samples are produced. However, I noticed that you change the default behavior of `tfd.MultivariateNormalDiag.sample`, which took me a whole night as I found my implementation yielded different results from yours. Could you please share your intension of this [code](https://github.com/danijar/dreamer/blob/02f0210f5991c7710826ca7881f19c64a012290c/tools.py#L417)?

I intend to move my question about the RSSM later as I think I may have found the answer(may be just a part of it). Again, I want to thank you -- I've learned a lot from your code!

**Comment by danijar** [comment]
The sampling ops are patched to ignore the seed. This is a performance optimization. Without seed, the sampling ops are faster because they do not need to be executed in sequence anymore. A simpler and more generic way to do this is the following:

```python
# Patch sampling to ignore seed to allow parallel execution.
_orig_random_categorical = tf.random.categorical
def random_categorical(*args, **kwargs):
  kwargs['seed'] = None
  return _orig_random_categorical(*args, **kwargs)
tf.random.categorical = random_categorical

# Patch sampling to ignore seed to allow parallel execution.
_orig_random_normal = tf.random.normal
def random_normal(*args, **kwargs):
  kwargs['seed'] = None
  return _orig_random_normal(*args, **kwargs)
tf.random.normal = random_normal
```

**Comment by yusukeurakami** [comment]
Sorry if I am not understanding your former answer correctly, but why we need  `SampleDist`? I especially don't understand what argmax in SampleDist.mode() is doing.

```python
tf.gather(sample, tf.argmax(logprob))[0]
```

**Comment by xlnwel** [comment]
Many `tfd.distributions` methods such as `entropy` become invalid after `tfd.TransformedDistribution` and `SampleDist` can be used to approximate those stats. The line you cited just find the sample with the maximum `logprob`.

**Comment by yusukeurakami** [comment]
@xlnwel  Thank you! I see.
